embedded-graphics/src/draw_target.rs

```diff
@@ -226,161 +221,66 @@ where
     /// [`core::convert::Infallible`]: https://doc.rust-lang.org/stable/core/convert/enum.Infallible.html
     type Error;
 
-    /// Draws a pixel on the display.
-    fn draw_pixel(&mut self, item: drawable::Pixel<C>) -> Result<(), Self::Error>;
-
-    /// Draws an object from an iterator over its pixels.
-    fn draw_iter<T>(&mut self, item: T) -> Result<(), Self::Error>
-    where
-        T: IntoIterator<Item = drawable::Pixel<C>>,
-    {
-        for pixel in item {
-            self.draw_pixel(pixel)?;
-        }
-
-        Ok(())
-    }
-
-    /// Returns the dimensions of the `DrawTarget` in pixels.
     fn size(&self) -> Size;
 
-    /// Clears the display with the supplied color.
+    // TODO: Mention performance
+    // TODO: Should this method take `Rectangle` instead of `&Rectangle`? The rest of e-g takes the former
+    // TODO: Mention default impl behaviour
+    /// Fill a given area with a transparent pixel iterator.
     ///
-    /// This default implementation can be replaced if the implementing driver provides an
-    /// accelerated clearing method.
-    fn clear(&mut self, color: C) -> Result<(), Self::Error>
+    /// Iteration order is guaranteed from top left to bottom right, however some pixels may be
+    /// transparent which are represented as `None`.
+    fn fill_sparse<I>(&mut self, area: &Rectangle, colors: I) -> Result<(), Self::Error>
     where
-        Self: Sized,
+        I: IntoIterator<Item = Option<Self::Color>>,
     {
-        primitives::Rectangle::new(Point::zero(), self.size())
-            .into_styled(PrimitiveStyle::with_fill(color))
-            .draw(self)
-    }
-
-    /// Draws a styled line primitive.
-    ///
-    /// This default trait method can be overridden if a display provides hardware-accelerated
-    /// methods for drawing lines.
-    ///
-    /// # Caution
-    ///
-    /// This method should not be called directly from application code. It is used to define the
-    /// internals of the [`draw`] method used for the [`Styled`] [`Line`] primitive. To draw a line,
-    /// call [`draw`] on a `Styled<Line>` object.
-    ///
-    /// [`Line`]: ../primitives/line/struct.Line.html
-    /// [`draw`]: ./trait.DrawTarget.html#method.draw
-    /// [`Styled`]: ../style/struct.Styled.html
-    fn draw_line(
-        &mut self,
-        item: &Styled<primitives::Line, PrimitiveStyle<C>>,
-    ) -> Result<(), Self::Error> {
-        self.draw_iter(item)
+        self.draw_iter(
+            area.points()
+                .zip(colors)
+                .filter_map(|(pos, color)| color.map(|c| Pixel(pos, c))),
+        )
```

Comment:
@rfuest 

This is obviously the incorrect behaviour, but how should we fix this for sparse primitives like `Circle`, etc? Should we add a wrapper iterator over styled primitive iterators to produce a bounding box iterator yielding `Option<C>`?

Reply:
> This is obviously the incorrect behaviour,

I think I'm a little slow today, how exactly is this the incorrect behavior?

I haven't really thought about how we should handle iterators in the future. If a drawable can be implemented by only calling functions in `DrawTarget`, like it's possible for rectangles, the iterator impl becomes kind of redundant. To keep the iterator support for primitives, which we should, we will have to maintain both variants in the future and make sure they are resulting in the same output.

> Should we add a wrapper iterator over styled primitive iterators to produce a bounding box iterator yielding `Option<C>`?

I think the other way around makes more sense in most cases. When a iterator, like for a styled circle, already iterates over the bounding box internally it would be easy to return a `Option<C>` instead of skipping these pixels.

Reply:
> I think I'm a little slow today, how exactly is this the incorrect behavior?

`area.points()` returns a contiguous iterator of x * y pixels, whereas `colors` will return fewer items as it's a sparse iterator.

> it would be easy to return a Option<C> instead of skipping these pixels.

It would indeed and I did consider this approach. Are we ok doing breaking changes to a primitive's styled iterator? They're pretty internal (albeit `pub`) so I don't think it'll break much/any user code.

Reply:
> `area.points()` returns a contiguous iterator of x * y pixels, whereas `colors` will return fewer items as it's a sparse iterator.

But `fill_sparse` takes an `Iterator<Item = Option<C>>` which is required to return x * y pixels. If `colors` would return fewer pixels there would be no way of knowing where to put the pixels, because the item only contains the color and not the position.

> It would indeed and I did consider this approach. Are we ok doing breaking changes to a primitive's styled iterator?

Yes, I expected that we would need to break the iterators during this refactoring.

Reply:
> But fill_sparse takes an Iterator<Item = Option<C>> which is required to return x * y pixels.

Sorry, I can see why my comment would be confusing now: when I made it, the current iterators don't do this, hence me mentioning the incorrect behaviour. At any rate, I'll change them to produce `Option<C>` :)

> Yes, I expected that we would need to break the iterators during this refactoring.

Sweet, I'll get that done asap :)